### DISCUSSION TITLE
How to select both File and Folder in StorageProvider?

### DISCUSSION THREAD

**Discussion by ghost** [AvaloniaUI/Avalonia#12771] [Q&A] [answered]
Hi, I would like to ask if Avalonia offers something for opening Dialog with both Folder and File selection, as I see Avalonia StorageProvider only provides selection of file or folder for one time.

Because my application should select both File and Folder at the same time, but pressing the file and then turning it back on and pressing the folder will cause a lot of trouble.

**Comment by maxkatz6** [comment, marked as answer]
There is no such thing in most platforms. IIRC only macOS allows selection of both.
We can't have an API that is only for a single platform.

If you target desktop only, you can take a look at ManagedFileChooser implementation and adjust it for your needs
https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Dialogs

It won't be a native picker, but since on desktop platforms you can have an access to the file system, you can build your own.

**Reply by maxkatz6** [reply]
Though, on macOS it might require special permissions.

**Reply by maxkatz6** [reply]
Alternatively, probably as a working fallback for other platforms, you can use Folder picker only, but then enumerate all storage items recursively with folder.GetItemsAsync(), and ask user to choose specific files or folders among them.

**Reply by timunie** [reply]
In addition one may use a managed dialog and provide this functionality on their own, given that this might not work for desktop only. Example: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Dialogs/ManagedFileChooser.cs
